### PR TITLE
Add example for metrics and dashboard

### PR DESCRIPTION
**Pull request by MarcoPolo**
This came up after a call with @Giulio2002 when I discussed they could set up these dashboards for some more observability and to help debug #2202 . 

Note for reviewer: Could you please try this and make sure it works for you as well (follow instructions in README.md)

### CODE REVIEW

File: examples/metrics-and-dashboards/README.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+metrics. Sets up a Prometheus server and Grafana server via Docker compose. A
+small go-libp2p dummy application is included to emit metrics.
+
+Run it with:
+
+```
+docker compose up
+```
+
+Go to http://localhost:3000/dashboards to see the dashboards.
```
Comment (lthibault) [inline examples/metrics-and-dashboards/README.md:13]: Missing newline at EOF

File: examples/metrics-and-dashboards/go-libp2p-node.Dockerfile [hunk 2]
```diff
@@ -0,0 +1,7 @@
+FROM golang:alpine
+WORKDIR /app
+COPY ./main.go .
+RUN go mod init example.com/m/v2
+RUN go mod tidy
+RUN go build main.go
+ENTRYPOINT [ "/app/main" ]
```
Comment (lthibault) [inline examples/metrics-and-dashboards/go-libp2p-node.Dockerfile:7]: Missing newline at EOF

File: examples/metrics-and-dashboards/grafana.Dockerfile [hunk 3]
```diff
@@ -0,0 +1,10 @@
+FROM grafana/grafana
+
+COPY ./dashboards /etc/grafana/dashboards
+USER root
+
+# Replace every instrance of ${DS_PROMETHEUS} with "prometheusuid"
+# Grafana doesn't let us do this in provisioning, so we do it ourselves.
+RUN sed -i 's/${DS_PROMETHEUS}/prometheusuid/g' /etc/grafana/dashboards/*/*.json 
+RUN cat /etc/grafana/dashboards/swarm/swarm.json
+USER grafana
```
Comment (lthibault) [inline examples/metrics-and-dashboards/grafana.Dockerfile:10]: Missing newline at EOF

**Review by lthibault — CHANGES_REQUESTED** [review]

**Review by sukunrt — APPROVED** [review]
LGTM! :tada: 
I've added minor comments. Feel free to ignore them if you think they're not relevant

File: examples/metrics-and-dashboards/main.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	if err != nil {
+		log.Fatal(err)
+	}
+	server, err := libp2p.New(libp2p.ResourceManager(rmgr))
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Make a bunch of clients that all ping the server at various times
+	wg := sync.WaitGroup{}
```
[the remark this answers is not in the document]
Comment (sukunrt) [reply examples/metrics-and-dashboards/main.go:45]: *server initiate a connection with some clients.

File: examples/metrics-and-dashboards/compose.yml [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+      context: ../../
+    environment:
+      - GF_AUTH_DISABLE_LOGIN_FORM=true
+      - GF_AUTH_ANONYMOUS_ENABLED=true
+      - GF_AUTH_ANONYMOUS_ORG_ROLE=Admin
+    volumes:
+      - ./grafana-config/grafana.ini:/etc/grafana/grafana.ini
+      - ./grafana-config/provisioning:/etc/grafana/provisioning
+    ports:
+      - 3000:3000
```
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply examples/metrics-and-dashboards/compose.yml:32]: I'm okay with that tradeoff

**Comment by MarcoPolo** [comment]
Deferring this work until after ipfs thing

**Comment by marten-seemann** [comment]
@MarcoPolo What's the status of this PR? It would be nice to get this merged before the next release.

**Comment by marten-seemann** [comment]
@MarcoPolo Friendly ping. Can we push this PR across the finish line?

**Comment by MarcoPolo** [comment]
Pushed changes, this is good to merge after CI passes.

**Comment by p-shahi** [comment]
> Pushed changes, this is good to merge after CI passes.

merged based on this message, mainly in service of https://github.com/libp2p/blog/pull/77

**Pull request merged**